rust/src/client/mobile_client/mod.rs

```diff
@@ -0,0 +1,70 @@
+use crate::{
+    client::Proxy,
+    crypto::{SecretSigningKey, SigningKeyPair},
+    mask::model::Model,
+};
+
+mod client;
+use self::client::ClientStateMachine;
+
+pub mod participant;
+use self::participant::ParticipantSettings;
+
+pub struct MobileClient {
+    runtime: tokio::runtime::Runtime,
+    client_state: Option<ClientStateMachine>,
```

Comment:
are we ever in a situation where the `client_state` is `None`? the `new()` and `next()` methods always guarantee a `Some(_)` and none of the other methods are `take()`ing this. the `Option<_>`s for the models can still be checked because they are optional within the state as well. hence this could simplify the code a bit.

Reply:
Yeah I tried to remove the optional, but I got some ownership issues in the `next` method because of the move of the client into the tokio runtime.

Reply:
For reference:

```rust
    pub fn next(&mut self) {
        if let Some(current_state) = self.client_state.take() {
            let new_state = self
                .runtime
                .block_on(async move { current_state.next().await });
            self.client_state = Some(new_state)
        }
    }
```

I don't see an alternative to `Option` here indeed. The only way around that would be for `next` signature to be changed to `fn next(self) -> Self` but that has ramifications in other places.

Reply:
i see, we would need to clone the state for the `async move`, which makes `take()` the better option. thanks for the explanation 👍

Reply:
Note that `block_on` doesn't require the future to be `Send + 'static`. The reason for using `move` here is that `ClientStateMachine::next` takes `self`. So we actually have two solutions if we want to remove that extra `Option`

- make `MobileClient::next` take `self` and return `Self` (as I suggested above)
- make `ClientStateMachine::next` take `&mut` and use `async` instead of `async move`

Regardless of the `Option` discussion, it would be nice if the `next` methods had the same signature btw. I'm just not sure how hard that would be.

Reply:
Do you mean with `next` methods the methods of each client variant (mobile, web, ..)?

Reply:
Well we have:

- `ClientStateMachine::next`: `async fn next(self) -> Self`
- `ClientState<T>::next`: `async fn next(self) -> ClientStateMachine`
- `MobileClient::next`: `fn next(&mut self)`
- `Participant<Sum>::next`: `fn next(self) -> Participant<Sum2>`

I'm saying it would be nice if these were a little more consistent, since they have the same name. In particular, making `MobileClient::next` an `fn next(self) -> Self` would be more consistent with `ClientStateMachine::next` and `ClientState<T>::next`, _and_ probably allow us to get rid of the `Option` @janpetschexain mentionned.

As a sidenote: `Participant<Sum>::next` could maybe be renamed into something like `into_sum2` or maybe we could `impl Into<Participant<Sum2>> for Participant<Sum>`? We're getting too many `next` methods.